packages/browser-runtime-core/src/open-context-runtime.ts

```diff
@@ -29,18 +30,17 @@ export class OpenContextRuntime implements Runtime {
       emit: (eventName: string, ...args: any[]) => void;
     }
   ) {
-    this.serviceProvider = serviceProvider;
     this.interpreterEnvironment = interpreterEnvironment;
-    this.shellEvaluator = new ShellEvaluator(serviceProvider, messageBus || new EventEmitter());
-    this.shellEvaluator.setCtx(this.interpreterEnvironment.getContextObject());
+    this.internalState = new ShellInternalState(serviceProvider, messageBus || new EventEmitter());
+    this.shellEvaluator = new ShellEvaluator(this.internalState);
+    this.internalState.setCtx(this.interpreterEnvironment.getContextObject());
     this.interpreter = new Interpreter(this.interpreterEnvironment);
   }
 
   async getCompletions(code: string): Promise<Completion[]> {
     if (!this.autocompleter) {
-      const buildInfo = await this.serviceProvider.buildInfo();
-      const serverVersion = buildInfo.version;
-      this.autocompleter = new ShellApiAutocompleter(serverVersion);
+      await this.internalState.fetchConnectionInfo();
+      this.autocompleter = new ShellApiAutocompleter(this.internalState.connectionInfo.buildInfo.version);
```

Comment:
This seems done in the same way in CLI repl, maybe we should keep it in mind and see if we can factor this somehow.

Reply:
Do you mean the autocompleter (which irina is currently pulling out) or do you mean the step of fetchConnectionInfo?